### PR TITLE
State example

### LINKED ISSUE #32: Add state example
The common extension points don't list state. I think it is an important example to add.

### PR DESCRIPTION
**Pull request by fcollonval**
Fix #32

### CODE REVIEW

**Comment by fcollonval** [comment]
@echarles I added an illustration. This is ready for review.

File: state/README.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
+# Using State Persistence in an Extension
+
+This example shows how to save and restore data save in persistent state database
```
Comment (jtpio) [inline state/README.md]: Small typo:

```suggestion
This example shows how to save and restore data saved in persistent state database
```

File: state/README.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+```ts
+// src/index.ts#L25-L27
+
+app.restored
+  // Get the state object
+  .then(() => state.fetch(PLUGIN_ID))
+```
+<!-- prettier-ignore-end -->
+
+The data are loaded as a `ReadonlyJSONValue` object. So all stored data must be
```
Comment (jtpio) [inline state/README.md]: I know it's usually debated, but I think the JupyterLab comments and docs use the singular for data:

```suggestion
The data is loaded as a `ReadonlyJSONValue` object. So all stored data must be
```

File: state/README.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+```ts
+// src/index.ts#L30-L32
+
+if (value) {
+  choice = (value as ReadonlyJSONObject)['choice'] as string;
+}
+```
+
+The `if` test ensure some value has been read. It is important to set a default value.
+Indeed the first time an user will installed your extension, the state won't contain
```
Comment (jtpio) [inline state/README.md]:
```suggestion
Indeed the first time a user will install the extension, the state won't contain
```

File: state/README.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+any value for your plugin.
```
Comment (jtpio) [inline state/README.md]:
```suggestion
any value for the plugin.
```

File: state/README.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+In the example, once the state is read, the user is prompted to choose a choice from
```
Comment (jtpio) [inline state/README.md]: How about?

```suggestion
In the example, once the state is read, the user is prompted to choose an option from
```

File: state/README.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+an item list with the default choice being stored as state variable.
```
Comment (jtpio) [inline state/README.md]: Missing `a`?

```suggestion
an item list with the default choice being stored as a state variable.
```

File: state/README.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+return InputDialog.getItem({
+  title: 'Pick a choice',
+  items: choices,
+  current: Math.max(0, choices.indexOf(choice))
+});
+```
+
+This implies to store the new choice done by the user in the state. This is done
+using the `save` methode of `IStateDB`:
```
Comment (jtpio) [inline state/README.md]:
```suggestion
using the `save` method of `IStateDB`:
```

File: state/README.md [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+```ts
+// src/index.ts#L47-L47
+
+return state.save(PLUGIN_ID, { choice });
+```
+
+To see it action,
```
Comment (jtpio) [inline state/README.md]: "In" action?

```suggestion
To see it in action,
```

File: state/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * Initialization data for the state extension.
+ */
+const extension: JupyterFrontEndPlugin<void> = {
+  id: PLUGIN_ID,
+  autoStart: true,
+  requires: [IStateDB],
+  activate: (app: JupyterFrontEnd, state: IStateDB) => {
+    const choices = ['one', 'two', 'three'];
```
Comment (jtpio) [inline state/src/index.ts]: How about naming it `options`?

**Comment by jtpio** [comment]
Thanks @fcollonval, this looks great!

(left a couple of comments)

**Comment by echarles** [comment]
@fcollonval This a very nice example. The picture does also a great jobs. @jtpio has been faster than me for the review and I agree that `option` is a better fit than `choice`.

To make it still more demonstrative, you could maybe change the label if the state exists (`two (from state)` instead of `two`), but this would just be icing on the cake.

**Comment by fcollonval** [comment]
Thank for correcting the English @jtpio 

I updated also the example 
- to bring the screencast at the README top 
- to add a link to JupyterLab dialog documentation.

**Pull request merged**